renderdoc/serialise/codecs/vk_cpp_codec/ext_object.h

```diff
@@ -0,0 +1,190 @@
+/******************************************************************************
+* The MIT License (MIT)
+*
+* Copyright (c) 2018 Baldur Karlsson
+*
+* Permission is hereby granted, free of charge, to any person obtaining a copy
+* of this software and associated documentation files (the "Software"), to deal
+* in the Software without restriction, including without limitation the rights
+* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+* copies of the Software, and to permit persons to whom the Software is
+* furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included in
+* all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+* THE SOFTWARE.
+******************************************************************************/
+#pragma once
+
+#include <assert.h>
+#include <math.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <array>
+#include <list>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "common/common.h"
+#include "core/core.h"
+#include "core/intervals.h"
+#include "serialise/rdcfile.h"
+
+namespace vk_cpp_codec
+{
+// Create a fake resource object.
+inline SDObject *makeResourceObject(const char *n, const char *t, uint64_t value)
```

Comment:
As mentioned before, `ResourceId` should be treated as an opaque object and not as a `uint64_t` which is an implementation detail. That also means that constructing an `SDObject` out of a `uint64_t` is not allowed. I can't suggest an alternative because I'm only seeing this function alone, but this will have to be refactored.

Reply:
I am not sure why this needs to be refactored? All this is meant to do is to create a 'fake' resource ID object. Here is a brief description how this is used right now, but there are other possible use cases:
1. The captured trace doesn't contain a vkAcquireNextImageKHR call, so the codec injects it manually. The vkAcquireNextImageKHR expects a semaphore, that the codec also generates manually.
2. Codec performs several preprocessor steps where it analyzes the entire trace for various aspects. It finds the command buffer that utilizes the acquired frame or it's derivative resources (images, image views or framebuffers) and inserts the wait semaphore into the first command buffer submission chunk as additional wait semaphore.
3. After that, the actual code generator just treats that chunk as a regular chunk. 

The actual resource ID that I use is checked to not collide with existing IDs at runtime.

Reply:
The issue is that it's constructing a `ResourceId` from an integer, and that's not allowed.

Your explanation makes sense and I can see why the exporter would need to insert extra calls. Adding them as chunks is reasonable to avoid too much special casing. I don't agree though that it necessarily follows that you need to be able to construct your own IDs from integers.

Instead I'd ask you to generate any resource IDs you need through the normal `ResourceIDGen::GetNewUniqueID()` which I can move out into more common code at some point (currently in `core/resource_manager.h` since it's only used by drivers at the moment). `ResourceId`s generated during capture - the ones that will be referenced in the structured data - start at 1 and increment atomically. During replay the `ResourceId` generation is offsetted to begin at 1000000000000000000, so that IDs generated at replay time never clash.

Reply:
>Instead I'd ask you to generate any resource IDs you need through the normal ResourceIDGen::GetNewUniqueID() 
That would be perfect, I actually wanted to use ResourceIDGen::GetNewUniqueID() when I wrote that code, but IIRC accessing it was restricted. 
>which I can move out into more common code at some point
How would you suggest untangling CL dependency? Are you OK to accept current implementation and have me change it once your CL lands or do you want me to wait until yours is done and rebase? Please let me know.

Reply:
I think if you `#include core/resource_manager.h` you *should* be able to access it from your code just as-is. I'd say just do that for now, and I can tidy it up to remove the dependency on that header sometime later. It's not high priority - still in the right project, just shouldn't be coupled like that so I can find a better home for it now that other things need to generate `ResourceId`s.

Reply:
Cool, will do that, thanks!